app/split_shares.go

```diff
@@ -0,0 +1,298 @@
+		Txs:      processedTxs,
+		Messages: messages,
+		Evidence: data.Evidence,
+	}
+}
+
+// shareSplitter write a data square using provided block data. It also ensures
+// that message and their corresponding txs get written to the square
+// atomically.
+type shareSplitter struct {
+	txWriter  *coretypes.ContiguousShareWriter
+	msgWriter *coretypes.MessageShareWriter
+
+	// we aren't doing anything with evidence or isrs yet, so we don't need to
+	// lazily generate those shares lazily
+	evdShares [][]byte
+	isrShares [][]byte
+
+	squareSize    uint64
+	maxShareCount int
+	txConf        client.TxConfig
+}
+
+func newShareSplitter(txConf client.TxConfig, squareSize uint64, data *core.Data) *shareSplitter {
+	sqwr := shareSplitter{
+		squareSize:    squareSize,
+		maxShareCount: int(squareSize * squareSize),
+		txConf:        txConf,
+	}
+
+	evdData := new(coretypes.EvidenceData)
+	err := evdData.FromProto(&data.Evidence)
+	if err != nil {
+		panic(err)
+	}
+	sqwr.evdShares = evdData.SplitIntoShares().RawShares()
+
+	sqwr.txWriter = coretypes.NewContiguousShareWriter(consts.TxNamespaceID)
+	sqwr.msgWriter = coretypes.NewMessageShareWriter()
+
+	return &sqwr
+}
+
+// writeTx marshals the tx and lazily writes it to the square. Returns true if
+// the write was successful, false if there was not enough room in the square.
+func (sqwr *shareSplitter) writeTx(tx []byte) (ok bool, err error) {
+	delimTx, err := coretypes.Tx(tx).MarshalDelimited()
+	if err != nil {
+		return false, err
+	}
+
+	if !sqwr.hasRoomForTx(delimTx) {
+		return false, nil
+	}
+
+	sqwr.txWriter.Write(delimTx)
+	return true, nil
+}
+
+// writeMalleatedTx malleates a MsgWirePayForData into a MsgPayForData and
+// its corresponding message provided that it has a MsgPayForData for the
+// preselected square size. Returns true if the write was successful, false if
+// there was not enough room in the square.
+func (sqwr *shareSplitter) writeMalleatedTx(
+	parentHash []byte,
+	tx signing.Tx,
+	wpfd *types.MsgWirePayForData,
+) (ok bool, malleatedTx coretypes.Tx, msg *core.Message, err error) {
+	// parse wire message and create a single message
+	coreMsg, unsignedPFD, sig, err := types.ProcessWirePayForData(wpfd, sqwr.squareSize)
+	if err != nil {
+		return false, nil, nil, err
+	}
+
+	// create the signed PayForData using the fees, gas limit, and sequence from
+	// the original transaction, along with the appropriate signature.
+	signedTx, err := types.BuildPayForDataTxFromWireTx(tx, sqwr.txConf.NewTxBuilder(), sig, unsignedPFD)
+	if err != nil {
+		return false, nil, nil, err
+	}
+
+	rawProcessedTx, err := sqwr.txConf.TxEncoder()(signedTx)
+	if err != nil {
+		return false, nil, nil, err
+	}
+
+	wrappedTx, err := coretypes.WrapMalleatedTx(parentHash[:], rawProcessedTx)
+	if err != nil {
+		return false, nil, nil, err
+	}
+
+	// check if we have room for both the tx and message it is crucial that we
+	// add both atomically, otherwise the block is invalid
```

Comment:
```suggestion
	// Check if we have room for both the tx and message. It is crucial that we
	// add both atomically, otherwise the block would be invalid.
```